### PR TITLE
querify layout and move param env out of the infcx

### PR DESCRIPTION
**Pull request by nikomatsakis**
The main goal of this PR is to move the parameter environment *out* of the inference context. This is because the inference environment will soon be changing over the course of inference --- for example, when we enter into a `for<'a> fn(...)` type, we will push a new environment with an increasing universe index, rather than skolemizing the `'a` references. Similarly, each obligation will soon be able to have a distinct parameter environment, and therefore the `Obligation` struct is extended to carry a `ParamEnv<'tcx>`. (I debated about putting it into the cause; seems plausible, but also weird.)

Along the way, I also reworked how layout works, moving the layout cache into a proper query along the lines of needs-drop and friends. 

Finally, tweaks the inference context API. It seemed to be accumulating parameters at an alarming rate. The main way to e.g. make a subtype or equality relationship is to do the following:

    infcx.at(cause, param_env).sub(a, b)
    infcx.at(cause, param_env).eq(a, b)

In both cases, `a` is considered the "expected" type (this used to be specified by a boolean). I tried hard to preserve the existing notion of what was "expected", although in some cases I'm not convinced it was being set on purpose one way or the other. This is why in some cases you will see me do `sup(b, a)`, which is otherwise equivalent to `sub(a, b)`, but sets the "expected type" differently.

r? @eddyb 
cc @arielb1

### CODE REVIEW

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -1690,6 +1690,219 @@ impl<'a, 'tcx> Layout {
...
+    fn record_layout_for_printing_outlined(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                           ty: Ty<'tcx>,
+                                           param_env: ty::ParamEnv<'tcx>,
+                                           layout: &Layout) {
+        // (delay format until we actually need it)
+        let record = |kind, opt_discr_size, variants| {
+            let type_desc = format!("{:?}", ty);
+            let overall_size = layout.size(tcx);
+            let align = layout.align(tcx);
+            tcx.sess.code_stats.borrow_mut().record_type_size(kind,
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: This is a bit of a hack but I can live with impure stats just like we have diagnostics.

**Comment by Mark-Simulacrum** [comment]
@bors r-

Should be a quick fix, this is just the UI tests updating due to no longer being counted:

```
[01:08:42] diff of stderr:
[01:08:42] 
[01:08:42]  error[E0391]: unsupported cyclic reference between types/traits detected
[01:08:42]    |
[01:08:42]  note: the cycle begins when computing layout of `S`...
[01:08:42]  note: ...which then requires computing layout of `std::option::Option<<S as Mirror>::It>`...
[01:08:42]  note: ...which then requires computing layout of `<S as Mirror>::It`...
[01:08:42]    = note: ...which then again requires computing layout of `S`, completing the cycle.
[01:08:42]  
[01:08:42] -error: aborting due to previous error
[01:08:42] +error: aborting due to previous error(s)
[01:08:42]  
[01:08:42] 
[01:08:42] The actual stderr differed from the expected stderr.
[01:08:42] 
[01:08:42] ------------------------------------------
[01:08:42] stderr:
[01:08:42] ------------------------------------------
[01:08:42] error[E0391]: unsupported cyclic reference between types/traits detected
[01:08:42]   |
[01:08:42] note: the cycle begins when computing layout of `S`...
[01:08:42] note: ...which then requires computing layout of `std::option::Option<<S as Mirror>::It>`...
[01:08:42] note: ...which then requires computing layout of `<S as Mirror>::It`...
[01:08:42]   = note: ...which then again requires computing layout of `S`, completing the cycle.
[01:08:42] 
[01:08:42] error: aborting due to previous error(s)
[01:08:42] 
[01:08:42] ------------------------------------------
[01:08:42] 
[01:08:42] thread '[ui] ui/issue-26548.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2472
[01:08:42] failures:
[01:08:42]     [ui] ui/issue-26548.rs
```

**Comment by Mark-Simulacrum** [comment]
@bors retry

 - Travis timed out after 6+ hours due to current problems, I hope this might indicate they've started to be resolved.

**Comment by nikomatsakis** [comment]
tfw you rebase something and, by the time you're done, there's another conflict

**Pull request merged**